Hosted plugins (#18)
bufbuild/docs.buf.build: docs/bsr/remote-generation/hosted-plugins.md

---
id: remote-plugin-execution
title: Remote Plugin Execution
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Remote Plugin Execution

> The remote plugin execution feature makes it possible to execute Protobuf plugins remotely in a secure environment on the BSR, as opposed to invoking plugins on your local machine.

Reviewer: I think we need a paragraph on why this is important (not having to manage your versions, why this is a pain, installation, etc)

Author: Expanded this section out.

By isolating and centralizing code generation from its environment, you eliminate an entire class of problems caused by subtle differences across specific versions of `protoc` and custom Protobuf plugins.

## Supported plugins

### `protoc`-based plugins

The Buf team has developed tooling to automatically sync and publish all of the plugins built-in to `protoc`, which are located under the `protocolbuffers` organization. A full list can be found here:

> https://buf.build/protocolbuffers/plugins

Reviewer: Expand the list here

This is powerful because you no longer need to have `protoc` installed. It provides all the benefits of code generation using `buf` and remote plugins without the headache of managing `protoc`.

### gRPC plugins

In addition to the plugins mentioned above, we're also adding support for popular gRPC plugins for nearly all of the same languages. These plugins are located under the `grpc` organization. A full list can be found here:

> https://buf.build/grpc/plugins

Reviewer: Expand the list

    out: gen/js
  - remote: buf.build/grpc/plugins/node:v1.11.2-1
    out: gen/js
```

  </TabItem>
  <TabItem value="python">

```yaml title=buf.gen.yaml
version: v1
managed:
  enabled: true
plugins:
  - remote: buf.build/protocolbuffers/plugins/python:v3.19.1-1
    out: gen/python
  - remote: buf.build/grpc/plugins/python:v1.41.1-1
    out: gen/python
```
  
  </TabItem>
  <TabItem value="ruby">

```yaml title=buf.gen.yaml
version: v1
managed:
  enabled: true
plugins:
  - remote: buf.build/protocolbuffers/plugins/ruby:v3.19.1-1
    out: gen/ruby
  - remote: buf.build/grpc/plugins/ruby:v1.41.1-1
    out: gen/ruby
```

  </TabItem>
  <TabItem value="java">

```yaml title=buf.gen.yaml
version: v1
managed:
  enabled: true
plugins:
  - remote: buf.build/protocolbuffers/plugins/java:v3.19.1-1
    out: gen/java
  - remote: buf.build/grpc/plugins/java:v1.42.1-1
    out: gen/java
```

  </TabItem>
</Tabs>

Note, we're using the `remote` key instead of `name` to reference a remote plugin, instead of a local one. More information can be [found in the buf.gen.yaml docs](https://docs.buf.build/configuration/v1/buf-gen-yaml#name-or-remote).

> > As a best practice, when referencing remote plugins we recommend including the version of the plugin to ensure reproducible code generation.

Reviewer 2: This note is nice, but I feel we're missing to mention beforehand that the version can be omitted. Maybe just above this line we can say something like:
```
A remote plugin can have a version specified, as is done above, or it can be omitted, if you want to always use the latest version of the plugin.
```
Then the note makes a lot of sense.

Author: Updated

It is possible to reference both local and remote plugins within a single template file. The `buf generate` command issues an RPC to the BSR to execute the remote plugins against the given module. Once execution is finished the output is written out to disk.

```terminal
$ buf generate buf.build/demolab/theweather
```

What you should end up with is the following structure:

<Tabs
  groupId="language-selection"
  defaultValue="go"
  values={[
    {label: 'Go', value: 'go'},
    {label: 'JavaScript', value: 'javascript'},
    {label: 'Python', value: 'python'},
    {label: 'Ruby', value: 'ruby'},
    {label: 'Java', value: 'java'},
  ]}>
  <TabItem value="go">

```bash
.
├── buf.gen.yaml
└── gen
    └── go
        └── weather
            └── v1
                ├── weather.pb.go
                └── weather_grpc.pb.go
```

  </TabItem>
  <TabItem value="javascript">

```bash
.
├── buf.gen.yaml
└── gen
    └── js
        ├── getweatherrequest.js
        ├── getweatherresponse.js
        └── weather
            └── v1

>                 └── weather_grpc_pb.js

Reviewer 2: I assume we tested this, but does this import the other files correctly?

```bash
.
├── buf.gen.yaml
└── gen
    └── python
        └── weather
            └── v1
                ├── weather_pb2.py
                └── weather_pb2_grpc.py
```
  
  </TabItem>
  <TabItem value="ruby">

```bash
.
├── buf.gen.yaml
└── gen
    └── ruby
        └── weather
            └── v1
                ├── weather_pb.rb
                └── weather_services_pb.rb
```

  </TabItem>
  <TabItem value="java">

```bash
.
├── buf.gen.yaml
└── gen
    └── java
        └── com
            └── weather
                └── v1
                    ├── GetWeatherRequest.java
                    ├── GetWeatherRequestOrBuilder.java
                    ├── GetWeatherResponse.java
                    ├── GetWeatherResponseOrBuilder.java
                    ├── WeatherProto.java
                    └── WeatherServiceGrpc.java
```

  </TabItem>
</Tabs>

## Wrapping up

Remote plugin execution simplifies the process of generating code for your Protobuf API. It also has the added benefit of enforcing reproducible outputs by eliminating differences in the environment where generation takes place, such as a developer's local machine or across continuous integration environments.

> All code generation takes place in a **secure environment** on the BSR.

Reviewer: Why is it secure?

Author: Chatted offline, removed this for now.

At Buf, we believe code generation is a key building block and the Protobuf ecosystem deserves a proper solution.

The **remote plugin execution** feature makes it possible to remotely generate source code using hosted plugins in an isolated environment on the BSR. By isolating code generation from its environment, you eliminate an entire class of problems caused by subtle differences across specific compiler versions and custom Protobuf plugins.

All you need to get started is:

- The `buf` CLI
- A [`buf.gen.yaml`](../../configuration/v1/buf-gen-yaml.md) file
- An [input](../../reference/inputs.md) of your choice

With this setup a single developer or thousands of developers at a large organization are able to achieve stable and reproducible code generation, while enjoying a simplified developer experience.

The Buf team has published a set of [Official plugins](#official-plugins) for you to use, starting with all the built-in `protoc` Protobuf plugins and popular ones such as gRPC plugins. 

- Interested in publishing your own community plugin? Check out [Authoring a Plugin](plugin-example.md)

> - To learn more about Buf Plugins check out the [Key concepts docs](concepts.md)

Author: Do you mean add a net new section and/or link for "best practices" or reword the existing "To learn more.." sentence?

> - To learn more about Buf Plugins check out the [Key concepts docs](concepts.md)

Reviewer 2: I mean to use `>` instead of `-`.

## Official plugins

### `protoc`-based plugins

The Buf team has developed tooling to automatically sync and publish all of the plugins built-in to `protoc`, which are located under the `protocolbuffers` organization. The following is a list of supported `protoc`-based plugins:

- https://buf.build/protocolbuffers/plugins/go
- https://buf.build/protocolbuffers/plugins/java
- https://buf.build/protocolbuffers/plugins/python
- https://buf.build/protocolbuffers/plugins/cpp
- https://buf.build/protocolbuffers/plugins/csharp
- https://buf.build/protocolbuffers/plugins/js
- https://buf.build/protocolbuffers/plugins/objc
- https://buf.build/protocolbuffers/plugins/php
- https://buf.build/protocolbuffers/plugins/ruby

> - https://buf.build/protocolbuffers/plugins/kotlin

Reviewer 2: This will eventually include the other plugins from https://github.com/bufbuild/remote-plugins/issues/7, could you leave a comment there to make sure we update this list after that is fixed?

In addition to the plugins mentioned above, we're also adding support for popular gRPC plugins for nearly all of the same languages. These plugins are located under the `grpc` organization. A full list can be found here:

https://buf.build/grpc/plugins

### Example

The following is an example of remote plugin execution, all you need is `buf`, a `buf.gen.yaml` file and a module of your choice.

We'll use the [https://buf.build/acme/petapis](acme/petapis) module hosted on the BSR. You can also use local Protobuf files, but for the sake of this example we're using a hosted module to illustrate remote plugin execution.

Next, create a `buf.gen.yaml` file with the following contents: 

```yaml title=buf.gen.yaml
version: v1
managed:
  enabled: true
  go_package_prefix:
    default: github.com/org/repo/gen/go
plugins:
  - remote: buf.build/protocolbuffers/plugins/go:v1.27.1-1
    out: gen/go
    opt: paths=source_relative
  - remote: buf.build/grpc/plugins/go:v1.1.0-1
    out: gen/go
    opt:
      - paths=source_relative
      - require_unimplemented_servers=false
```

And then run the following command:

```terminal
$ buf generate buf.build/acme/petapis --include-imports
```

> What this does is send the `acme/petapis` module up to the BSR, resolve dependencies (because we included the `--include-imports` flag), and then execute the requested plugins in a secure sandbox environment on the BSR. Once execution is finished the output is written out to local disk.

Author: This isn't entirely accurate, need to reword this.

---
id: remote-plugin-execution
title: Remote Plugin Execution
---

# Remote Plugin Execution

The remote plugin execution feature makes it possible to execute Protobuf plugins remotely in a secure environment on the BSR, as opposed to invoking plugins on your local machine.

By isolating and centralizing code generation from its environment, you eliminate an entire class of problems caused by subtle differences across specific versions of `protoc` and custom Protobuf plugins.

## Supported plugins

### `protoc`-based plugins

The Buf team has developed tooling to automatically sync and publish all of the plugins built-in to `protoc`, which are located under the `protocolbuffers` organization. A full list can be found here:

https://buf.build/protocolbuffers/plugins

> This is powerful because you no longer need to have `protoc` installed. All the benefits of code generation using `buf` and remote plugins without the headache of managing `protoc`.

Reviewer 2: This second sentence felt weird, what about
```suggestion
This is powerful because you no longer need to have `protoc` installed. It provides all the benefits of code generation using `buf` and remote plugins without the headache of managing `protoc`.
```

### gRPC plugins

In addition to the plugins mentioned above, we're also adding support for popular gRPC plugins for nearly all of the same languages. These plugins are located under the `grpc` organization. A full list can be found here:

https://buf.build/grpc/plugins

### Example

> The following is an example of remote plugin execution, all you need to get started:

Reviewer 2: This also reads a bit stuttery to me, how about
```suggestion
The following is an example of remote plugin execution, and this is all you need to get started:
```

- `buf`
- A [`buf.gen.yaml`](../../configuration/v1/buf-gen-yaml.md) file 
- An [input](../../reference/inputs.md) of your choice

We'll use the [buf.build/demolab/theweather](https://buf.build/demolab/theweather) module hosted on the BSR as the input source. You can also use local Protobuf files, but for this example we'll use a hosted module to illustrate remote plugin execution.

Create a template file with the following contents: 

```yaml title=buf.gen.yaml
version: v1
managed:
  enabled: true
  go_package_prefix:
    default: github.com/org/repo/gen/go
plugins:
  - remote: buf.build/protocolbuffers/plugins/go:v1.27.1-1
    out: gen/go
    opt: paths=source_relative
  - remote: buf.build/grpc/plugins/go:v1.1.0-1
    out: gen/go
    opt:
      - paths=source_relative

>       - require_unimplemented_servers=false

Reviewer 2: It might be cool to use a language other than Go, since this is already kinda easy in Go. A language like Python or Ruby, for example, is already super hard to use with `buf`: https://github.com/bufbuild/buf/issues/223 https://github.com/bufbuild/buf/issues/370

Reviewer 2: Also, we should reply on those issues with examples of how to do this today 🚀

Reviewer 2: I replied on the issues

Author: Do you think we should have examples for a few/all of the languages? Something like this..

![CleanShot 2021-11-16 at 10 32 25@2x](https://user-images.githubusercontent.com/6278244/142015416-944b7e9d-6e0d-4c36-9b44-4ac376aef4cf.png)

I worry about adding too many examples as it'll make the page too verbose.

Reviewer 2: It would be a super cool use of the tabs feature, maybe?

Author: Brilliant idea, let me try that.

```

> Note, we're using the `remote` key instead of `name` to reference a remote plugin, instead of a local one. More information can be [found here](https://docs.buf.build/configuration/v1/buf-gen-yaml#name-or-remote).

Reviewer 2: Nit: I prefer to be explicit about where the link goes. What do you think?
```suggestion
Note, we're using the `remote` key instead of `name` to reference a remote plugin, instead of a local one. More information can be [found in the buf.gen.yaml docs](https://docs.buf.build/configuration/v1/buf-gen-yaml#name-or-remote).
```

> As a best practice, when referencing remote plugins we recommend including the version of the plugin to ensure reproducible code generation.

It is possible to reference both local and remote plugins within a single template. The `buf generate` command issues an RPC to the BSR to execute the remote plugins against the given module. Once execution is finished the output is written out to disk.

```terminal
$ buf generate buf.build/demolab/theweather
```

What you should end up with is the following structure:

```bash
.
├── buf.gen.yaml
└── gen
    └── go
        └── weather
            └── v1
                ├── weather.pb.go
                └── weather_grpc.pb.go
```

## Wrapping up

Remote plugin execution simplifies the process of generating code for your Protobuf API. It also has the added benefit of enforcing reproducible outputs by eliminating differences in the environment where generation takes place, such as a developer's local machine or across continuous integration environments.

All code generation takes place in a **secure environment** on the BSR.

> Bring your own Protobuf files, or publish them to the BSR, and then generate the corresponding client and server code in your language of choice with hosted plugins on the BSR. You get all the benefits of code generation without the headache of managing plugins.

Reviewer 2: ```suggestion
Bring your own Protobuf files, or publish them to the BSR, and then generate the corresponding client and server code in your language of choice with hosted plugins on the BSR. You get all the benefits of code generation without the headache of managing plugins or `protoc` versions.
```

---
id: remote-plugin-execution
title: Remote Plugin Execution
---

# Remote Plugin Execution

The remote plugin execution feature makes it possible to execute Protobuf plugins remotely in a secure environment on the BSR, as opposed to invoking plugins on your local machine.

By isolating and centralizing code generation from its environment, you eliminate an entire class of problems caused by subtle differences across specific versions of `protoc` and custom Protobuf plugins.

> As a best practice, when referencing remote plugins we recommend including the version of the plugin to ensure reproducible code generation.

Reviewer 3: We might want to move this line closer to the example with the `remote: ` configuration below - otherwise it might not hit home since the user doesn't know what we're talking about yet.

## Supported plugins

### `protoc`-based plugins

The Buf team has developed tooling to automatically sync and publish all of the plugins built-in to `protoc`, which are located under the `protocolbuffers` organization. A full list can be found here:

https://buf.build/protocolbuffers/plugins

This is powerful because you no longer need to have `protoc` installed. All the benefits of code generation using `buf` and remote plugins without the headache of managing `protoc`.

### gRPC plugins

In addition to the plugins mentioned above, we're also adding support for popular gRPC plugins for nearly all of the same languages. These plugins are located under the `grpc` organization. A full list can be found here:

https://buf.build/grpc/plugins

### Example

The following is an example of remote plugin execution, all you need to get started:

- `buf`

> - A `buf.gen.yaml` file 

Reviewer 3: Let's make this a link to the `buf.gen.yaml` configuration reference.

> - A module of your choice

Reviewer 3: We might want to say `An input` and link to the `input` reference page.

> We'll use the [acme/petapis](https://buf.build/acme/petapis) module hosted on the BSR as the input source. You can also use local Protobuf files, but for this example we'll use a hosted module to illustrate remote plugin execution.

Reviewer 3: Although verbose, let's include the `buf.build/` prefix in the format so that it's immediately clear that we're talking about a module hosted on `buf.build`.

Create a template file with the following contents: 

```yaml title=buf.gen.yaml
version: v1
managed:
  enabled: true
  go_package_prefix:
    default: github.com/org/repo/gen/go
plugins:
  - remote: buf.build/protocolbuffers/plugins/go:v1.27.1-1
    out: gen/go
    opt: paths=source_relative
  - remote: buf.build/grpc/plugins/go:v1.1.0-1
    out: gen/go
    opt:
      - paths=source_relative
      - require_unimplemented_servers=false
```

> Note, we're using the `remote` key instead of `name` to reference a remote plugin, instead of a local one. More information can be [found here](https://docs.buf.build/configuration/v1/buf-gen-yaml#name-or-remote). 

Reviewer 3: This is where I'd include a note syntax (`> `) immediately below this line and reference the recommendation to use the versioned suffix (the point you have above).

It is possible to reference both local and remote plugins within a single template.

> The `buf generate` command issues an RPC to the BSR to execute the remote plugins against the given module. Once execution is finished the output is written out to disk.

Reviewer 3: Depending on how appealing this is rendered on the page, consider combining these two sentences into one since they're somewhat related and it would remove some whitespace. It might be better as-is though - just food for thought.

```terminal
$ buf generate buf.build/acme/petapis --include-imports
```

> The acme/petapis module has a dependency on the acme/paymentapis and googleapis/googleapis so we include the `--include-imports` flag to resolve those dependencies.

Reviewer 3: ```suggestion
The acme/petapis module has a dependency on the acme/paymentapis and googleapis/googleapis so we include the `--include-imports` flag to also generate code for these dependencies.
```

However, note that we probably don't want to generate `buf.build/googleapis/googleapis` in this case. In the tour, we recommend that users set the `managed.go_package_prefix.except` value so that the Go types are imported from `genproto`.

We might want to come up with a different example, or require that the user runs both `buf generate buf.build/acme/petapis` and `buf generate buf.build/acme/paymentapis`.

Author: Ye, I've gone back and forth on this. For the sake of simplicity we should use a different example that has no dependency on `buf.build/googleapis/googleapis` or requires a multiple generate calls.

I think it's outside the scope of "remote plugin execution" to get into the nuances of this generation. The goal is to illustrate a simple but working example of remote plugin execution.

What you should end up with is the following structure:

```bash
.
├── buf.gen.yaml
└── gen
    └── go
        ├── google
        │   └── type
        │       ├── datetime.pb.go
        │       └── money.pb.go
        ├── payment
        │   └── v1alpha1
        │       └── payment.pb.go
        ├── pet
        │   └── v1
        │       ├── pet.pb.go
        │       └── pet_grpc.pb.go
        └── weather
            └── v1
                ├── weather.pb.go
                └── weather_grpc.pb.go

> ```

Reviewer 3: I think you're already planning this, but would be great to have something that concludes / ties everything together. Doesn't need to be a formal `## Conclusion` section, but a couple sentences would be good.
